use GetTransactions method for fetching tickets

oshorefueled opened the pull request:
This PR replaces the use of getTicket method with GetTransactions to fetch wallet tickets. It uses different transaction filters to fetch tickets with different status.
Fix #546
Fix #580
Fix #570 
Fix #474 
Close #458

The ticket activity page and section on the overview page are commented out. The code implemented looks like a duplicate of the ticket list page and having it as part of the ticket module is redundant. I suggest that page be reviewed before it is added back as an accessible feature.

oshorefueled committed:
clean up wallet/commands.go

- remove unused methods
- remove multiwallet methods that can be called directly from pages since
  pages already have multiwallet in the Load struct.

oshorefueled committed:
add ticket type and methods to load package

oshorefueled committed:
clean up ticket list page

- add ticket methods for fetch different ticket types in the load package

oshorefueled committed:
resolve inactive button bug on ticket purchase modal

oshorefueled committed:
remove global ticket implementation

oshorefueled committed:
cleanup code

beansgum commented on ui/load/ticket.go:
```diff
@@ -0,0 +1,169 @@
+	StakingVoted    = "VOTED"
+)
+
+func (wl *WalletLoad) StakingOverviewAllWallets() *dcrlibwallet.StakingOverview {
```
This should be done in dcrlibwallet instead.

beansgum commented on ui/page/tickets/overview.go:
```diff
@@ -165,109 +174,125 @@ func (pg *Page) ticketPriceSection(gtx layout.Context) layout.Dimensions {
-}
-
-func (pg *Page) stackingRecordSection(gtx layout.Context) layout.Dimensions {
+//func (pg *Page) ticketsActivitySection(gtx layout.Context) layout.Dimensions {
```
Remove unused code comments

oshorefueled replied:
This is the activity page that is currently commented out which I suggested be removed before it is added back

beansgum commented on ui/load/ticket.go:
```diff
@@ -0,0 +1,169 @@
+	var tickets []Ticket
+	wallets := wl.MultiWallet.AllWallets()
+
+	liveTicketFilters := []int32{dcrlibwallet.TxFilterImmature, dcrlibwallet.TxFilterLive}
```
Why are immature tickets being fetched with live tickets?

oshorefueled replied:
For the overview page. The overview page displays both immature and live tickets in the live ticket section
![Screenshot 2021-08-20 at 14 04 30](https://user-images.githubusercontent.com/15804688/130237537-42abe9aa-f199-4353-9d00-b10cc1df08c7.png)

beansgum commented on ui/load/ticket.go:
```diff
@@ -0,0 +1,169 @@
+			}
+			tickets = append(tickets, transactionsToTickets(tx, filterToStatus(filter), w.Name)...)
+		}
+	}
```
This block can be replaced with this https://github.com/planetdecred/dcrlibwallet/blob/master/transactions.go#L118

beansgum commented on ui/load/ticket.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+func (wl *WalletLoad) GetTickets(walletID int, txFilter int32, newestFirst bool) ([]Ticket, error) {
+	if txFilter == dcrlibwallet.TxFilterStaking {
```
Why not pass the staking filter directly to the `w.GetTransactionsRaw` function?

beansgum commented on ui/load/ticket.go:
```diff
@@ -0,0 +1,169 @@
+		if err != nil {
+			return nil, err
+		}
+	}
```
Why not?
```go
txs, err := w.GetTransactionsRaw(0, 0, dcrlibwallet.TxFilterStaking, newestFirst)
if err != nil {
    return nil, err
}
```

oshorefueled replied:
currently the ticket status such as immature and expired tickets can't be gotten directly from the transaction without calculation. Except for voted and revoked tickets which are set as transaction types. 
the calculations are already done in dcrlibwallet, I feel its redundant to calculate it again in godcr

beansgum commented on ui/load/ticket.go:
```diff
@@ -0,0 +1,169 @@
+		overview.Voted += ov.Voted
+	}
+
+	wl.MultiWallet.GetLowestBlockTimestamp()
```
What does this do?

beansgum reviewed (dismissed):
- The tickets page freezes the app a bit before opening. 
- Some of the tooltips added by https://github.com/planetdecred/godcr/pull/488 aren't showing

beansgum commented on ui/load/ticket.go:
```diff
@@ -0,0 +1,169 @@
+	}
+}
+
+func filterToStatus(txFilter int32) string {
```
It looks like dcrlibwallet tx filters can be referenced directly instead of converting to identical constants.

beansgum commented on ui/page/tickets/list_page.go:
```diff
@@ -71,11 +69,66 @@ func (pg *ListPage) ID() string {
+		txFilter = dcrlibwallet.TxFilterExpired
+	case "Revoked":
+		txFilter = dcrlibwallet.TxFilterRevoked
+	default:
```
Use `pg.ticketTypeDropDown.SelectedIndex()` because the value of `Selected()` can change if another language is used.

beansgum commented on ui/page/tickets/overview.go:
```diff
@@ -165,109 +174,125 @@ func (pg *Page) ticketPriceSection(gtx layout.Context) layout.Dimensions {
+		{"LIVE", pg.stakingOverview.Live},
+		{"VOTED", pg.stakingOverview.Voted},
+		{"EXPIRED", pg.stakingOverview.Expired},
+		{"REVOKED", pg.stakingOverview.Revoked},
```
Use dcrlibwallet tx filter constants.

oshorefueled committed:
use dcrlibwallet staking PR

oshorefueled committed:
cleanup fetch ticket code

oshorefueled committed:
delete ticket file from load package

oshorefueled committed:
move tooltip initialization to onResume

beansgum committed:
Cleanup tickets overview page

- display correct tickets reward value

beansgum committed:
Fix only tickets purchase tx showing as 'All' in tickets list page

- add transactionItem to wrap ticket transaction and tooltips

beansgum committed:
Add shadow to LinearLayout

- Apply shadow to more list items and back up seed bottom pane

beansgum committed:
Add GridLayout

- use GridLayout for tickets grid view to fix scrolling lag issues

beansgum committed:
tickets: fix ticket cards layouts

- show ticket maturity progress and display correct value on time label
- use correct progress bar and track colors
- add TxStatus struct
- sort ticket transactions and prioritize spent tickets

beansgum committed:
Fix ticket status tooltip

- add tooltip messages for all ticket statuses and format block time
  for voted and revoked tickets
- fix bug where overview page shows the same details for all tickets
- add ticket status and confirmations property to transactionItem struct
- add shadow to tooltip and remove border stroke

beansgum committed:
Fix content displayed in ticket age, days to vote and maturity duration tooltips

beansgum committed:
Use transactionItem to populate tickets list values

- Add second progress bar layout that fixes a border radius bug
- Add shared data to transactionItem
- Refresh window after loading data to remove lag in ui update
- Use correct format for ticket maturity time

beansgum committed:
Fix popups and complete tx details not showing on tx overview page

beansgum committed:
Remove tickets activity page and unused util functions

beansgum committed:
Cleanup ticket purchase modal

Sirmorrison commented on ui/page/tickets/list_page.go:
```diff
@@ -251,35 +327,26 @@ func (pg *ListPage) ticketListLayout(gtx layout.Context, tickets []wallet.Ticket
 									r := func(gtx C) layout.Dimensions {
-										txt := pg.Theme.Label(values.TextSize14, tickets[index].DaysBehind)
+
+										if ticket.ticketAge == "" {
```
```suggestion
										if components.StringNotEmpty(ticket.ticketAge) {
```

Sirmorrison commented on ui/page/tickets/purchase_modal.go:
```diff
@@ -60,6 +55,22 @@ func newTicketPurchaseModal(l *load.Load) *ticketPurchaseModal {
+	tp.vspSelector = newVSPSelector(tp.Load).title("Select a vsp")
+	tp.ticketPrice = dcrutil.Amount(tp.WL.TicketPrice())
+
+	if tp.vspIsFetched && tp.WL.GetRememberVSP() != "" {
```
```suggestion
	if tp.vspIsFetched && components.StringNotEmpty(tp.WL.GetRememberVSP()) {
```

Sirmorrison commented on ui/page/tickets/utils.go:
```diff
@@ -3,150 +3,243 @@ package tickets
+
+			ticketAgeDuration := time.Since(time.Unix(tx.Timestamp, 0)).Seconds()
+			ticketAge = ticketAgeTimeFormat(int(ticketAgeDuration))
+
```
Remove whitespace

Sirmorrison commented on ui/page/tickets/utils.go:
```diff
@@ -3,150 +3,243 @@ package tickets
 		layout.Rigid(func(gtx C) D {
-			if mainMsgDesc != "" {
-				toolTipContent(layout.Inset{Top: values.MarginPadding8}, mainMsgLabel2.Layout)
+			if mainDesc != "" {
```
```suggestion
			if components.StringNotEmpty(mainDesc) {
```

Sirmorrison commented on ui/page/tickets/utils.go:
```diff
@@ -3,150 +3,243 @@ package tickets
+			return D{}
+		}),
+		layout.Rigid(func(gtx C) D {
+			if subDesc != "" {
```
```suggestion
			if components.StringNotEmpty(subDesc) {
```

Sirmorrison commented on ui/page/tickets/utils.go:
```diff
@@ -457,6 +504,26 @@ func createClickGestures(count int) []*gesture.Click {
+	}
+
+	return fmt.Sprintf("%ds", secs)
+
```
remove whitespace

Sirmorrison commented on ui/page/tickets/utils.go:
```diff
@@ -457,6 +504,26 @@ func createClickGestures(count int) []*gesture.Click {
 	return gestures
 }
 
+func ticketAgeTimeFormat(secs int) string {
```
This method is an exact replica of `nextTicketRemaining` Line 527. Only difference is that `nextTicketRemaining` takes into account weeks.

beansgum replied:
Actually, this method returns the time in a single rounded off unit("1d", "1h", or "5m") but nextTicketRemaining returns the complete time remaining ("1h20m5s").

Sirmorrison commented on ui/values/dimensions.go:
```diff
@@ -4,6 +4,7 @@ import "gioui.org/unit"
 
 var (
 	Size0_5 = unit.Dp(0.5)
+	Size28  = unit.Dp(28)
```
Why don't we have this as the regular `MarginPadding28` and Size0_5 as `MarginPadding0p5`

beansgum replied:
I'm slowly migrating to using `Size` as the prefix instead of `MargingPadding` as the values are not only used for margins and paddings.

Sirmorrison commented on ui/window.go:
```diff
@@ -291,7 +288,7 @@ func (win *Window) Loop(w *app.Window, shutdown chan int) {
 			}
 
-			win.updateStates(e.Resp)
+			// win.updateStates(e.Resp)
```
Remove if no longer needed.

beansgum committed:
Resolve requested changes

beansgum committed:
Show error if selected ticket purchase account has insufficient balance

Sirmorrison reviewed (changes requested):
* Any specific reason the ticket activity page was removed?

* When you have only one ticket, the ticket is center aligned rather than being aligned to the top left corner of the screen.
* 
![image](https://user-images.githubusercontent.com/27733432/133417639-dbc9b91e-0ec0-4ed3-9c82-b0aeb46f861f.png)

* if I click the purchase button without entering my spending password (N:B i needed a password for my wallet), it says Attempting to Purchase ticket (1) but no tickets are purchased.

I understand that we should be able to allow for an empty password but since I currently need a password and non was provided, which also did not allow the ticket to be purchase (i think so), I should get an error message of some sort stating that the ticket purchase was not successful.

![image](https://user-images.githubusercontent.com/27733432/133423739-1bc7497c-dcc5-4636-9580-a1024ead9d8b.png)

* Button color is not app default blue color. 
![image](https://user-images.githubusercontent.com/27733432/133424066-8184e516-486e-4022-bda4-f0b7bf980f75.png)

beansgum committed:
Process ticket purchase in review modal

Sirmorrison reviewed (changes requested):
After successfully purchasing as ticket, the staking record still shows `0 unmined ` tickets instead of 1.
![image](https://user-images.githubusercontent.com/27733432/133449599-c9b98cbf-2ab4-4283-94ff-a9669d5ed1f5.png)

![image](https://user-images.githubusercontent.com/27733432/133449678-b0542c79-2c1c-49d6-9c15-1705aebce5db.png)

Waiting a while, i see that i now have one unmined ticket as a live ticket, and one staking record however, in the view all ticket page, the ticket is shown as immature.

overview page

![image](https://user-images.githubusercontent.com/27733432/133450561-22c73d13-f13b-4d18-a0a8-acd6ffa7c1d3.png)

All tickets page
![image](https://user-images.githubusercontent.com/27733432/133450680-ff4edfb3-5a3f-4f75-a663-935b01d6f55e.png)

SO to narrow down the issues, the ticket overview page is only refreshed when i Navigate to another Page and then back to the ticket page.

I believe at this point, the OnResume() is called which updates the page.

beansgum committed:
Refresh tickets overview data after successful ticket purchase

Sirmorrison reviewed (commented):
Pr looks good with functionality.

Some observations.
* On the all tickets page, the wallet name is not displayed on the ticket card.
* When i hover on the wallet name from the ticket overview page, there is no tooltip pop up. This was earlier implemented inPR #488.

Ticket overview page.
<img width="258" alt="image" src="https://user-images.githubusercontent.com/27733432/133476466-64ade621-4411-4c4f-9e60-3945790de0d1.png">

all ticket page
<img width="503" alt="image" src="https://user-images.githubusercontent.com/27733432/133476367-66b085e2-6a58-4470-941f-cc501553c9c2.png">

This is the mock up
<img width="407" alt="image" src="https://user-images.githubusercontent.com/27733432/133476249-d8c70149-ea6e-4e8b-bcb7-c88772907630.png">

beansgum committed:
Add tooltip to wallet name

beansgum committed:
Use dcrlibwallet master(31878a6)
